gh-115178: Add Counts of UOp Pairs to pystats

JeffersGlass opened the pull request:
Adds the ability to track pairs, triples, and longer sequences of UOps to the `pystats` statistics.

Currently, this only counts these statistics for the ***non-JIT tier 2 interpreter***. I intend to include tracking in the JIT in a follow-up PR.

To try it out, run:

```sh
./configure --enable-pystats
./make
mkdir /tmp/py_stats
PYTHON_UOPS=1 ./python -X pystats -c $'x = 0\nfor i in range(1000):\n x+=i'
```

The output file at `/tmp/py_stats/???.txt` will contain a new section with UOp Sequence counts:
```
UOp sequence count[BUILD_TUPLE,_CHECK_VALIDITY]: 2
UOp sequence count[LIST_APPEND,_CHECK_VALIDITY]: 13
UOp sequence count[LOAD_DEREF,_SET_IP]: 2
UOp sequence count[LOAD_FAST,LOAD_FAST]: 7
...
```

`Tools/scripts/summarize_stats.py` has also been expanded to encompass these new stats, with a new section:

<details>
<summary> Counts for top 100 UOp Sequences of Length 2 </summary>

|Sequence | Count | Self | Cumulative | 
|---|---:|---:|---:|
| _CHECK_VALIDITY,_SET_IP | 3,932 | 19.0% | 19.0% |
| LOAD_NAME,_CHECK_VALIDITY | 1,966 | 9.5% | 28.6% |
| _SET_IP,LOAD_NAME | 1,966 | 9.5% | 38.1% |
| _ITER_CHECK_RANGE,_SET_IP | 984 | 4.8% | 42.9% |
| _GUARD_NOT_EXHAUSTED_RANGE,_ITER_CHECK_RANGE | 984 | 4.8% | 47.6% |
| STORE_NAME,_SET_IP | 983 | 4.8% | 52.4% |
| STORE_NAME,_CHECK_VALIDITY | 983 | 4.8% | 57.1% |
| _SET_IP,STORE_NAME | 983 | 4.8% | 61.9% |
| _SET_IP,_BINARY_OP_ADD_INT | 983 | 4.8% | 66.7% |
| _SET_IP,_ITER_NEXT_RANGE | 983 | 4.8% | 71.4% |
| _SET_IP,_JUMP_TO_TOP | 983 | 4.8% | 76.2% |
| _GUARD_BOTH_INT,_CHECK_VALIDITY | 983 | 4.8% | 81.0% |
| _BINARY_OP_ADD_INT,_GUARD_BOTH_INT | 983 | 4.8% | 85.7% |
| _ITER_NEXT_RANGE,_GUARD_NOT_EXHAUSTED_RANGE | 983 | 4.8% | 90.5% |
| _JUMP_TO_TOP,_CHECK_VALIDITY | 983 | 4.8% | 95.2% |
| _CHECK_VALIDITY,STORE_NAME | 983 | 4.8% | 100.0% |
</details>

Finally, a new environment variable, `PYTHONSTATS_UOPDEPTH` can be set to any integer to record sequences of length longer than two. For example:

```sh
PYTHONSTATS_UOPDEPTH=4 PYTHON_UOPS=1 ./python -X pystats -c 'x = 0; for i in range(1000): x+=i
```

Yields:

<details>
<summary> Counts for top 100 UOp Sequences of Length 2 </summary>

|Sequence | Count | Self | Cumulative | 
|---|---:|---:|---:|
| _CHECK_VALIDITY,_SET_IP | 3,932 | 19.0% | 19.0% |
| LOAD_NAME,_CHECK_VALIDITY | 1,966 | 9.5% | 28.6% |
| _SET_IP,LOAD_NAME | 1,966 | 9.5% | 38.1% |
| _ITER_CHECK_RANGE,_SET_IP | 984 | 4.8% | 42.9% |
| _GUARD_NOT_EXHAUSTED_RANGE,_ITER_CHECK_RANGE | 984 | 4.8% | 47.6% |
| STORE_NAME,_SET_IP | 983 | 4.8% | 52.4% |
| STORE_NAME,_CHECK_VALIDITY | 983 | 4.8% | 57.1% |
| _SET_IP,STORE_NAME | 983 | 4.8% | 61.9% |
| _SET_IP,_BINARY_OP_ADD_INT | 983 | 4.8% | 66.7% |
| _SET_IP,_ITER_NEXT_RANGE | 983 | 4.8% | 71.4% |
| _SET_IP,_JUMP_TO_TOP | 983 | 4.8% | 76.2% |
| _GUARD_BOTH_INT,_CHECK_VALIDITY | 983 | 4.8% | 81.0% |
| _BINARY_OP_ADD_INT,_GUARD_BOTH_INT | 983 | 4.8% | 85.7% |
| _ITER_NEXT_RANGE,_GUARD_NOT_EXHAUSTED_RANGE | 983 | 4.8% | 90.5% |
| _JUMP_TO_TOP,_CHECK_VALIDITY | 983 | 4.8% | 95.2% |
| _CHECK_VALIDITY,STORE_NAME | 983 | 4.8% | 100.0% |


</details>

<details>
<summary> Counts for top 100 UOp Sequences of Length 3 </summary>

|Sequence | Count | Self | Cumulative | 
|---|---:|---:|---:|
| LOAD_NAME,_CHECK_VALIDITY,_SET_IP | 1,966 | 9.5% | 9.5% |
| _SET_IP,LOAD_NAME,_CHECK_VALIDITY | 1,966 | 9.5% | 19.0% |
| _CHECK_VALIDITY,_SET_IP,LOAD_NAME | 1,966 | 9.5% | 28.6% |
| _GUARD_NOT_EXHAUSTED_RANGE,_ITER_CHECK_RANGE,_SET_IP | 984 | 4.8% | 33.3% |
| STORE_NAME,_SET_IP,_BINARY_OP_ADD_INT | 983 | 4.8% | 38.1% |
| STORE_NAME,_CHECK_VALIDITY,_SET_IP | 983 | 4.8% | 42.9% |
| _SET_IP,STORE_NAME,_CHECK_VALIDITY | 983 | 4.8% | 47.6% |
| _SET_IP,_BINARY_OP_ADD_INT,_GUARD_BOTH_INT | 983 | 4.8% | 52.4% |
| _SET_IP,_ITER_NEXT_RANGE,_GUARD_NOT_EXHAUSTED_RANGE | 983 | 4.8% | 57.1% |
| _SET_IP,_JUMP_TO_TOP,_CHECK_VALIDITY | 983 | 4.8% | 61.9% |
| _GUARD_BOTH_INT,_CHECK_VALIDITY,_SET_IP | 983 | 4.8% | 66.7% |
| _BINARY_OP_ADD_INT,_GUARD_BOTH_INT,_CHECK_VALIDITY | 983 | 4.8% | 71.4% |
| _ITER_CHECK_RANGE,_SET_IP,_JUMP_TO_TOP | 983 | 4.8% | 76.2% |
| _ITER_NEXT_RANGE,_GUARD_NOT_EXHAUSTED_RANGE,_ITER_CHECK_RANGE | 983 | 4.8% | 80.9% |
| _JUMP_TO_TOP,_CHECK_VALIDITY,STORE_NAME | 983 | 4.8% | 85.7% |
| _CHECK_VALIDITY,STORE_NAME,_SET_IP | 983 | 4.8% | 90.5% |
| _CHECK_VALIDITY,_SET_IP,STORE_NAME | 983 | 4.8% | 95.2% |
| _CHECK_VALIDITY,_SET_IP,_ITER_NEXT_RANGE | 983 | 4.8% | 100.0% |


</details>

<details>
<summary> Counts for top 100 UOp Sequences of Length 4 </summary>

|Sequence | Count | Self | Cumulative | 
|---|---:|---:|---:|
| _SET_IP,LOAD_NAME,_CHECK_VALIDITY,_SET_IP | 1,966 | 9.5% | 9.5% |
| _CHECK_VALIDITY,_SET_IP,LOAD_NAME,_CHECK_VALIDITY | 1,966 | 9.5% | 19.0% |
| LOAD_NAME,_CHECK_VALIDITY,_SET_IP,LOAD_NAME | 983 | 4.8% | 23.8% |
| LOAD_NAME,_CHECK_VALIDITY,_SET_IP,STORE_NAME | 983 | 4.8% | 28.6% |
| STORE_NAME,_SET_IP,_BINARY_OP_ADD_INT,_GUARD_BOTH_INT | 983 | 4.8% | 33.3% |
| STORE_NAME,_CHECK_VALIDITY,_SET_IP,_ITER_NEXT_RANGE | 983 | 4.8% | 38.1% |
| _SET_IP,STORE_NAME,_CHECK_VALIDITY,_SET_IP | 983 | 4.8% | 42.9% |
| _SET_IP,_BINARY_OP_ADD_INT,_GUARD_BOTH_INT,_CHECK_VALIDITY | 983 | 4.8% | 47.6% |
| _SET_IP,_ITER_NEXT_RANGE,_GUARD_NOT_EXHAUSTED_RANGE,_ITER_CHECK_RANGE | 983 | 4.8% | 52.4% |
| _SET_IP,_JUMP_TO_TOP,_CHECK_VALIDITY,STORE_NAME | 983 | 4.8% | 57.1% |
| _GUARD_BOTH_INT,_CHECK_VALIDITY,_SET_IP,LOAD_NAME | 983 | 4.8% | 61.9% |
| _BINARY_OP_ADD_INT,_GUARD_BOTH_INT,_CHECK_VALIDITY,_SET_IP | 983 | 4.8% | 66.7% |
| _ITER_CHECK_RANGE,_SET_IP,_JUMP_TO_TOP,_CHECK_VALIDITY | 983 | 4.8% | 71.4% |
| _GUARD_NOT_EXHAUSTED_RANGE,_ITER_CHECK_RANGE,_SET_IP,_JUMP_TO_TOP | 983 | 4.8% | 76.2% |
| _ITER_NEXT_RANGE,_GUARD_NOT_EXHAUSTED_RANGE,_ITER_CHECK_RANGE,_SET_IP | 983 | 4.8% | 80.9% |
| _JUMP_TO_TOP,_CHECK_VALIDITY,STORE_NAME,_SET_IP | 983 | 4.8% | 85.7% |
| _CHECK_VALIDITY,STORE_NAME,_SET_IP,_BINARY_OP_ADD_INT | 983 | 4.8% | 90.5% |
| _CHECK_VALIDITY,_SET_IP,STORE_NAME,_CHECK_VALIDITY | 983 | 4.8% | 95.2% |
| _CHECK_VALIDITY,_SET_IP,_ITER_NEXT_RANGE,_GUARD_NOT_EXHAUSTED_RANGE | 983 | 4.8% | 100.0% |


</details>

<!-- gh-issue-number: gh-115178 -->
* Issue: gh-115178
<!-- /gh-issue-number -->

EDIT: I see this has automatically requested a review from a huge number of people. This is my first time submitting a PR to CPython - please do forgive me if I've goofed this process up somehow.

<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--115181.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

JeffersGlass committed:
Adjust structure of optimization_stats->opcode

Make optimization_stats.opcode an array of pointers to UOpStats objects,
instead of a member array.
Add a *next_stats[512] field to the UOpStats struct
Add _init_pystats function to initialize these new member structs
Call _init_pystats inside _PyConfig_Write

JeffersGlass committed:
Output data from uop chains

Add print_uop_sequence function to specialize.c to output stored
chains of uops

JeffersGlass committed:
Outputing sequences of UOps as Numbers!

JeffersGlass committed:
Output sequence chains by name

JeffersGlass committed:
Add missing check for _Py_stats

JeffersGlass committed:
Adjust formatting on sequence output

JeffersGlass committed:
Add uop sequence pairs to info

JeffersGlass committed:
Fix multi-length sequence output

JeffersGlass committed:
Output top 100 sequences of all present lengths

JeffersGlass committed:
Move initialization to check in '_stats_on'

JeffersGlass committed:
Add PYTHONSTATS_UOPDEPTH environment variable

Sets the maximum depth of UOP sequences to track. Defaults to 2.

JeffersGlass committed:
Catch up with main

JeffersGlass committed:
Catch up with main again

JeffersGlass committed:
Formatting cleanup

JeffersGlass committed:
Add docs, ACK

JeffersGlass committed:
Add docs, ACK

erlend-aasland committed:
Merge branch 'main' into uop-sequence-count

mdboom commented on Python/specialize.c:
```diff
@@ -292,6 +350,15 @@ void
 _Py_StatsOn(void)
 {
     _Py_stats = &_Py_stats_struct;
+    if (!_Py_stats->optimization_stats.opcode[0]) _init_pystats();
```
It's typical (especially when collecting stats from from the pyperformance suite) to turn stats on and off multiple times during a run, for example only around the "code of interest".  I think that this will reset and realloc the stats structures every time it's turned on, and probably leak the old memory, too.

mdboom commented on Tools/scripts/summarize_stats.py:
```diff
@@ -312,6 +312,16 @@ def get_opcode_stats(self, prefix: str) -> OpcodeStats:
+    def get_uop_sequence_stats(self, length) -> dict[str, int]:
+        return {k: v for k, v in self._data.items()
+                 if k.startswith("UOp sequence count") and
+                 k.count(',') == length - 1}
```
Nit: This file follows black formatting rules (though CI doesn't enforce it).

JeffersGlass committed:
Merge branch 'uop-sequence-count' of https://github.com/jeffersglass/cpython into uop-sequence-count

JeffersGlass committed:
Address linting errors

JeffersGlass committed:
Minor formatting, table renaming in summarize_stats.py

mdboom commented on Python/ceval.c:
```diff
@@ -1026,6 +1026,7 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
         next_uop++;
         OPT_STAT_INC(uops_executed);
         UOP_STAT_INC(uopcode, execution_count);
+        UOP_CHAIN_UPDATE(uopcode);
```
I think this may create invalid chains across trace runs.  For example, the last opcode in the last trace will be "remembered" and notated as if it precedes the first opcode of the next trace.  We should probably "reset the chain" at the start of every loop a few lines above here.

mdboom replied:
Also, since the `last_opcodes` array is in global memory (the main pystats struct), it's not thread safe.  If we ever had executors running in parallel they would step on each other.  It's probably better to have that in a local variable in this function.

JeffersGlass replied:
I think the latest commit (switching to just pair counting) resolves this by moving the `last_opcode` variable inside `_PyEval_EvalFrameDefault` - does it seem like my new approach is correct?

mdboom commented on Python/specialize.c:
```diff
@@ -221,6 +247,28 @@ print_histogram(FILE *out, const char *name, uint64_t hist[_Py_UOP_HIST_SIZE])
+// UOp with the given index
+static void
+print_uop_sequence(FILE *out, UOpStats *uop_stats, const char* prefix){
+    for (int i = 1; i < 512; i++){
```
Why 1 and not 0 here?

mdboom commented on Python/specialize.c:
```diff
@@ -42,6 +43,31 @@ PyStats *_Py_stats = NULL;
+        UOpStats *opcode_stat = _Py_stats_struct.optimization_stats.opcode[i];
+        opcode_stat = PyMem_RawCalloc(1, sizeof(UOpStats));
+        for (int j = 0; j < 512; j++){
+            opcode_stat->next_stats[j] = NULL;
```
This probably isn't necessary because `Calloc` is used above (which would already fill the buffer with zeros).

mdboom reviewed (commented):
Thanks for this PR.  It's definitely valuable work.

I can't help but think that the dynamic chain length support adds a lot of complexity that could bite us, and maybe we don't need it.  I'll let @brandtbucher speak to whether 3-length chains (and beyond) would even be feasible in the JIT.  But it sure would significantly simplify this PR to only handle 2-length chains.

Beyond just the complexity of having dynamically allocated buffers that are tricky to handle and free in the right places, we would probably need to use some sort of sparse representation for 4-length chains. This current approach would use 1TB of RAM.  (512**4 * 16)

mdboom commented on Include/internal/pycore_code.h:
```diff
@@ -314,6 +337,7 @@ PyAPI_FUNC(PyObject*) _Py_GetSpecializationStats(void);
 #define UOP_STAT_INC(opname, name) ((void)0)
 #define OPT_UNSUPPORTED_OPCODE(opname) ((void)0)
 #define OPT_HIST(length, name) ((void)0)
+#define UOP_CHAIN_UPDATE(op) printf("UOP CHAINER NOT RUNNING\n");
```
```suggestion
#define UOP_CHAIN_UPDATE(op) ((void)0)
```

JeffersGlass committed:
Remove my 'formatting fix' that was causing segfaults

JeffersGlass committed:
Implement pair counts

JeffersGlass committed:
Format python with Black

JeffersGlass committed:
Rework UOP_PAIR_INC to be no-op in non-stats case

JeffersGlass committed:
Add check for _Py_stats to avoid segfault

JeffersGlass committed:
Undo some of Black's formatting

JeffersGlass committed:
Merge branch 'main' into uop-sequence-count

JeffersGlass committed:
Track UOp pairs instead of lengthy sequences

This is a full rewrite of this PR - gone is the dynamic allocation, the tree of
UOp nodes, etc. Now were just track adjacent pairs of UOps in each (Non-jit)
executor. This is, as predicated, much simpler.

JeffersGlass committed:
Swap lastuop/current uop in counts

mdboom commented on Include/internal/pycore_code.h:
```diff
@@ -286,6 +286,9 @@ extern int _PyStaticCode_Init(PyCodeObject *co);
 #define UOP_STAT_INC(opname, name) do { if (_Py_stats) { assert(opname < 512); _Py_stats->optimization_stats.opcode[opname].name++; } } while (0)
+#define UOP_PAIR_INC(uopcode, lastuop) do { if (lastuop && _Py_stats) { \
+    _Py_stats->optimization_stats.opcode[lastuop].pair_count[uopcode]++; } \
+    lastuop = uopcode; } while (0)
```
I know it's a bit annoying with the `\` continues, but can you make the line breaks look more like what PEP7 suggests here?

JeffersGlass replied:
For sure - I've adjusted that now, mostly aping the example in PEP7 macros.

mdboom reviewed (approved):
This looks great!  (Modulo the re-formatting request).  Thanks for this contribution.

JeffersGlass committed:
Catch up with main

JeffersGlass committed:
Adjust formatting to better match PEP 7

JeffersGlass committed:
Revert "Catch up with main"

This reverts commit 03db7a5e950639cbbaea38ad4521f55ab39b1495.

JeffersGlass committed:
Merge branch 'main' into uop-sequence-count

markshannon commented on Include/cpython/pystats.h:
```diff
@@ -99,6 +99,7 @@ typedef struct _gc_stats {
 typedef struct _uop_stats {
     uint64_t execution_count;
     uint64_t miss;
+    uint64_t pair_count[512];
```
Can you replace 512 with `MAX_UOP_ID+1`.

markshannon commented on Python/specialize.c:
```diff
@@ -270,6 +270,15 @@ print_optimization_stats(FILE *out, OptimizationStats *stats)
         }
     }
+
+    for (int i = 0; i < 512; i++){
```
`0` is reserved and is the initial value of `lastuop` and 512 should be replaced with `MAX_UOP_ID+1`
```suggestion
    for (int i = 1; i <= MAX_UOP_ID; i++){
```

markshannon commented on Python/specialize.c:
```diff
@@ -270,6 +270,15 @@ print_optimization_stats(FILE *out, OptimizationStats *stats)
     }
+
+    for (int i = 0; i < 512; i++){
+        for (int j = 0; j < 512; j++) {
```
```suggestion
        for (int j = 1; j <= MAX_UOP_ID; j++) {
```

markshannon reviewed (changes requested):
A couple of minor issues, but looks good in general.

JeffersGlass committed:
Merge branch 'main' into uop-sequence-count

JeffersGlass committed:
Address review comments

JeffersGlass committed:
Catch up with main

JeffersGlass committed:
Undo spurious changes to prefixes

JeffersGlass committed:
Add title keyword arg to pair count tables

JeffersGlass committed:
Catch up with main

markshannon reviewed (approved):
Thanks @JeffersGlass for doing this.
This is going to be very useful for selecting micro-op pairs for the JIT.
